### PR TITLE
[WIP] modules

### PR DESCRIPTION
**Pull request by frangarcj**
This is some work on load modules, a fix for double relocations (code2) and some random implementations to get Downwell to run.

Feel free to pick it and continue the work

### CODE REVIEW

File: src/emulator/modules/SceAppMgr/SceAppMgrUser.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ EXPORT(int, sceAppMgrAppParamGetString, int pid, int param, char *string, int le
     switch (param) {
     case 12:
```
Comment (Rinnegatamante) [inline src/emulator/modules/SceAppMgr/SceAppMgrUser.cpp]: Probably would be better to have an enum for those.

File: src/emulator/kernel/src/load_self.cpp [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    for (size_t i = 0; i < count; ++i) {
+        const uint32_t nid = nids[i];
+        const Ptr<uint32_t> entry = entries[i];
+        
+        if (LOG_IMPORTS) {
+            const char *const name = import_name(nid);
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+        }
+        
+        /*uint32_t *const stub = entry.get(mem);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: ?

File: src/emulator/kernel/src/load_self.cpp [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    const sce_module_imports_raw *const imports_begin = reinterpret_cast<const sce_module_imports_raw *>(base + module.import_top);
+    const sce_module_imports_raw *const imports_end = reinterpret_cast<const sce_module_imports_raw *>(base + module.import_end);
+
+    for (const sce_module_imports_raw *imports = imports_begin; imports < imports_end; imports = reinterpret_cast<const sce_module_imports_raw *>(reinterpret_cast<const uint8_t *>(imports) + imports->size)) {
+        if (LOG_IMPORTS) {
+            const char *const lib_name = Ptr<const char>(imports->module_name).get(mem);
+            LOG_INFO("Loading imports from {}", lib_name);
+        }
+
+        //assert(imports->version == 1);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: why was this commented? please don't comment asserts if not completely sure, and when you do just remove the line

File: src/emulator/kernel/src/load_self.cpp [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        //assert(imports->num_syms_vars == 0);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: same

File: src/emulator/kernel/src/load_self.cpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+            continue;
+        
+        kernel.export_nids.emplace(nid,entry.address());
+        
+        if (LOG_EXPORTS) {
+            const char *const name = import_name(nid);
+            
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+        }
+        /*uint32_t *const stub = entry.get(mem);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: ?

File: src/emulator/kernel/src/load_self.cpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+            
+        }
+        
+        /*if (nid == 0x935cd196) {
+            entry_point = entry;
+            return true;
+        }*/
+        
+        kernel.export_nids.emplace(nid,entry.address());
+        /*uint32_t *const stub = entry.get(mem);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: ?

File: src/emulator/kernel/src/load_self.cpp [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        }
+        
+        if (LOG_EXPORTS) {
+            const char *const name = import_name(nid);
+            
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+            
+        }
+        
+        /*if (nid == 0x935cd196) {
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: Please add a comment explaining any commented code, why it's there, why it's commented etc.

File: src/emulator/kernel/src/load_self.cpp [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+static bool load_exports(Ptr<const void> &entry_point, const sce_module_info_raw &module, Ptr<const void> segment_address, KernelState &kernel, const MemState &mem) {
+    const uint8_t *const base = segment_address.cast<const uint8_t>().get(mem);
+    const sce_module_exports_raw *const exports_begin = reinterpret_cast<const sce_module_exports_raw *>(base + module.export_top);
+    const sce_module_exports_raw *const exports_end = reinterpret_cast<const sce_module_exports_raw *>(base + module.export_end);
+
+    for (const sce_module_exports_raw *exports = exports_begin; exports < exports_end; exports = reinterpret_cast<const sce_module_exports_raw *>(reinterpret_cast<const uint8_t *>(exports) + exports->size)) {
+        const char *const lib_name = Ptr<const char>(exports->module_name).get(mem);
+        //if(!lib_name)
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: ?

File: src/emulator/kernel/src/load_self.cpp [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+    return true;
+}
+
+SceUID load_self(Ptr<const void> &entry_point, KernelState &kernel, MemState &mem, const void *self, const char *path) {
+    const uint8_t *const self_bytes = static_cast<const uint8_t *>(self);
+    const SCE_header &self_header = *static_cast<const SCE_header *>(self);
+
+    // assumes little endian host
+    // TODO: do it in a better way, perhaps with user-defined literals that do the conversion automatically (magic != "SCE\0"_u32)
```
Comment (VelocityRa) [inline src/emulator/kernel/src/load_self.cpp]: this was my comment, just delete it

File: src/emulator/kernel/src/relocation.cpp [hunk 10]
```diff
@@ -212,7 +212,7 @@ bool relocate(const void *entries, size_t size, const SegmentAddresses &segments
             entry = short_entry + 1;
         } else {
             const LongEntry *const long_entry = static_cast<const LongEntry *>(entry);
-            assert(long_entry->code2 == 0);
+            //assert(long_entry->code2 == 0);
```
Comment (VelocityRa) [inline src/emulator/kernel/src/relocation.cpp]: afaik this line can be removed completely

File: src/emulator/modules/SceCtrl/SceCtrl.cpp [hunk 11]
```diff
@@ -222,30 +245,11 @@ EXPORT(int, sceCtrlPeekBufferNegative2) {
 }
 
 EXPORT(int, sceCtrlPeekBufferPositive, int port, SceCtrlData *pad_data, int count) {
-    assert(port == 0);
+    /*assert(port == 0);
```
Comment (VelocityRa) [inline src/emulator/modules/SceCtrl/SceCtrl.cpp]: why comment these asserts?

**Comment by VelocityRa** [comment]
As said in Discord, formatting issues will be resolved with #141 after merging this.

File: src/emulator/load_self.cpp [hunk 12]
```diff
@@ -1,160 +0,0 @@
-// Vita3K emulator project
```
Comment (VelocityRa) [inline src/emulator/load_self.cpp]: Moving this file, and making changes to it should be in separate commits IMO, it's impossible to see what you changed this way.

File: src/emulator/modules/SceGxm/SceGxm.cpp [hunk 13]
```diff
@@ -1049,16 +1060,55 @@ EXPORT(int, sceGxmSetFragmentTexture, SceGxmContext *context, unsigned int textu
         constexpr auto A_max = std::numeric_limits<decltype(A)>::max();
 
         switch (texture::get_swizzle(fmt)) {
-        case SCE_GXM_TEXTURE_SWIZZLE4_ABGR: R = 0; G = 1; B = 2; A = 3; break;
```
Comment (VelocityRa) [inline src/emulator/modules/SceGxm/SceGxm.cpp]: As I've already told you on Discord (just leaving this here so you remember it), you need to remove this diff and if you want, add the `// clang ...` stuff. Else just leave it as is (do change it back to the compact version though) and Rin will add those 2 lines in reformatting PR.

File: src/emulator/vpk.cpp [hunk 14]
```diff
@@ -45,31 +46,59 @@ static size_t write_to_buffer(void *pOpaque, mz_uint64 file_ofs, const void *pBu
...
-    std::memset(zip.get(), 0, sizeof(*zip));
-
-    FILE* vpk_fp;
-
-#ifdef WIN32
-    if (_wfopen_s(&vpk_fp, path.c_str(), L"rb")) {
-#else
-    if (!(vpk_fp = fopen(wide_to_utf(path).c_str(), "rb"))) {
-#endif
+static bool read_file_from_zip(Buffer &eboot, FILE *&vpk_fp, const char * file, const ZipPtr zip) {
```
Comment (VelocityRa) [inline src/emulator/vpk.cpp]: Make first arg name more generic. For example eboot -> buffer.

File: src/emulator/vpk.cpp [hunk 15]
```diff
@@ -45,31 +46,59 @@ static size_t write_to_buffer(void *pOpaque, mz_uint64 file_ofs, const void *pBu
...
+    if (!mz_zip_reader_init_cfile(zip.get(), vpk_fp, 0, 0)) {
         return false;
     }
 
-    if (!mz_zip_reader_init_cfile(zip.get(), vpk_fp, 0, 0)) {
+    const int eboot_index = mz_zip_reader_locate_file(zip.get(), file, nullptr, 0);
```
Comment (VelocityRa) [inline src/emulator/vpk.cpp]: Same.

**Comment by frangarcj** [comment]
Cause of #154
